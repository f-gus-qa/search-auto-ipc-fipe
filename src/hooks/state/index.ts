export { default as useAutoState } from './useAutoState';
export { default as useLoadingState } from './useLoadingState';
export { default as useSelectState } from './useSelectState';
export { default as useAlertState } from './useAlertState';
