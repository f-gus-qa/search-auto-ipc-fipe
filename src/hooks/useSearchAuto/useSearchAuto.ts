import { useSelectState } from '../state';
import useAppActions from '../useAppActions';
import useFipeClient from '../useFipeClient';

function useSearchAuto() {
  const fipeClient = useFipeClient();
  const { brand, model, year } = useSelectState();
  const {
    setSelected,
    startLoading,
    stopLoading,
    setAlert,
  } = useAppActions();

  const searchAuto = async () => {
    if (!brand || !model || !year) {
      return;
    }

    startLoading('all');

    try {
      setSelected('value', await fipeClient.getValueFromAuto({
        brandId: brand.id,
        modelId: model.id,
        yearId: year.id,
      }));
    } catch (e) {
      setAlert({
        type: 'error',
        message:
          'Não foi possível acessar o valor da tabela fipe, por favor tente novamente mais tarde',
      });
    } finally {
      stopLoading('all');
    }
  };

  return searchAuto;
}

export default useSearchAuto;
