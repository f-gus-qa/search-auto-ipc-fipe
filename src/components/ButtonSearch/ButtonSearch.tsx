import { Button, Grid, makeStyles } from '@material-ui/core';
import { omit } from 'lodash';
import { useSelectState } from '@/hooks/state';
import useSearchAuto from '@/hooks/useSearchAuto';

const useStyles = makeStyles((theme) => ({
  conatinerButton: {
    padding: `${theme.spacing(3)}px 0`,
    display: 'flex',
    justifyContent: 'center',
  },
}));

function ButtonSearch() {
  const selectedAuto = useSelectState();
  const classes = useStyles();
  const enableToSearch = !Object.values(omit(selectedAuto, ['value'])).includes(null);
  const searchAuto = useSearchAuto();

  return (
    <Grid item xs={12} className={classes.conatinerButton}>
      <Button
        data-testid="button-search"
        size="large"
        color="primary"
        variant="contained"
        onClick={searchAuto}
        disabled={!enableToSearch}
      >
        Consultar Pre√ßo
      </Button>
    </Grid>
  );
}

export default ButtonSearch;
