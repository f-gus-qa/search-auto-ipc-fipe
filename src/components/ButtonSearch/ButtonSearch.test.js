import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import brandFixture from '@/fixture/brand';
import modelFixture from '@/fixture/model';
import yearFixture from '@/fixture/year';
import valueFixture from '@/fixture/value';

import useSearchAuto from '@/hooks/useSearchAuto';
import { useSelectState } from '@/hooks/state';
import ButtonSearch from './ButtonSearch';

jest.mock('@/hooks/state', () => ({
  useSelectState: jest.fn(),
}));

useSelectState.mockReturnValue({
  brand: brandFixture(),
  model: modelFixture(),
  year: yearFixture(),
  value: valueFixture(),
});

jest.mock('@/hooks/useSearchAuto', () => jest.fn());

describe('<ButtonSearch>', () => {
  it('check renders', () => {
    const { toJSON } = create(<ButtonSearch />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('must call useSearchAuto', () => {
    render(<ButtonSearch />);

    const button = screen.queryByTestId('button-search');
    fireEvent.click(button);

    expect(useSearchAuto).toHaveBeenCalledTimes(1);
  });

  describe('when an selected value is not setted', () => {
    useSelectState.mockReturnValue({
      brand: brandFixture(),
      model: modelFixture(),
      year: null,
      value: valueFixture(),
    });

    it('check Button is disabled', () => {
      render(<ButtonSearch />);

      const button = screen.queryByTestId('button-search');
      expect(button).toHaveAttribute('disabled');
    });
  });
});
