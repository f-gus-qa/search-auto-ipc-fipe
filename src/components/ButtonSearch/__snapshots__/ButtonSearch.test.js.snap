// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonSearch> check renders 1`] = `
<div
  className="MuiGrid-root makeStyles-conatinerButton-1 MuiGrid-item MuiGrid-grid-xs-12"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled Mui-disabled"
    data-testid="button-search"
    disabled={true}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Consultar Pre√ßo
    </span>
  </button>
</div>
`;
